### PR TITLE
obtaining capability in worker [in progress]

### PR DESCRIPTION
**Pull request by evan-schott**
I'm confused if I'm going down the wrong path here or not. Logically doesn't makes sense that need to query to get capability because you should already have it locally (the gateway has already exported it right). But on the other hand couldn't really find any other way to go from node.Vat to getting a capability.

### CODE REVIEW

**Comment by evan-schott** [comment]
@lthibault

**Comment by evan-schott** [comment]
Also I think this is all I need to do for bullet point 3 of #2

File: main.go [hunk 1]
```diff
@@ -160,12 +160,20 @@ func SlightEchoHandler(writer http.ResponseWriter, request *http.Request) {
...
+	if err == nil {
+		return err
+	}
+	n.Vat.Connect(c.Context, g)
+
 	return errors.New("WORKER NOT IMPLEMENTED")
 }
 
-func waitPeers(c *cli.Context, n *server.Node) (bool, error) {
+func waitPeers(c *cli.Context, n *server.Node) (peer.ID, error) {
```
Comment (lthibault) [inline main.go:195]: ~I think this change is where you're getting thrown off-track. Remember that you're running several instances of this program, in separate processes.  Thus, each process only needs to make the decision "am _I_ the gateway, or is someone else?".~

If you are not the gateway (i.e. you are a worker), creating the capability in the worker is a simple matter of setting up the echo capnp server, and then creating a client for it (just like in the tutorial).  You don't need to export it or do anything fancy.  Ultimately you'll be sending it down the channel.

Comment (lthibault) [reply main.go:195]: Oh, I think I see where you're going:   you're passing the gateway's `peer.ID` to the worker, so it knows where to get the channel.  👍 

Keep in mind that there are two capabilities at play:

1. The channel (owned and exported by the gateway)
2. Each worker's `Echo` client (owned by the worker; **not** exported, but will be sent down the gateway's channel)

Does this help?

File: main.go [hunk 2]
```diff
@@ -160,12 +160,20 @@ func SlightEchoHandler(writer http.ResponseWriter, request *http.Request) {
 	writer.Write(responsePayload)
 }
 
-func runWorker(c *cli.Context, n *server.Node) error {
+func runWorker(c *cli.Context, n *server.Node, g peer.ID) error { // TODO: Is there a reason all these have just c,n as arguments?
 	time.Sleep(10 * time.Second)
+	var f, release = n.Cluster.View().Lookup(c.Context) // TODO: Submit a query for ("gatewayID/lb/chan")?
```
Comment (lthibault) [inline main.go]: You already know the gateway's ID now.  So what you should do is `n.Vat.Connect(...)` to get the channel from the gateway.

Pro tip:  when connecting, don't worry about the gateway peer's address.  If you just pass in `&peer.AddrInfo{ID: g}`, wetware will use the DHT to figure out the rest.

Comment (evan-schott) [reply main.go]: I am confused what I should put in as the 3rd argument of n.Vat.Connect()

Comment (evan-schott) [reply main.go]: Like the first two seem straightforward, but I'm not sure how I get the capability to put into the 3rd slot.

File: main.go (hunk 1 above)
Comment (evan-schott) [reply main.go:195]: Yeah, the part I am confused about is how the worker references the channel capability when trying to do n.Vat.Connect(). (Sorry for the spam)

File: main.go (hunk 2 above)
Comment (lthibault) [reply main.go]: See [here](https://pkg.go.dev/github.com/wetware/casm/pkg#Capability).  Remember last time we met, how we dug into how capabilities were mapped onto libp2p streams?  

In your case, you'll want to use [`BasicCap`](https://pkg.go.dev/github.com/wetware/casm/pkg#BasicCap) to define which stream will service your channel.  If you grep for `BasicCap`, you should see some straightforward examples.

Comment (evan-schott) [reply main.go]: Ok thank you!

File: main.go [hunk 3]
```diff
@@ -160,12 +160,20 @@ func SlightEchoHandler(writer http.ResponseWriter, request *http.Request) {
...
+	defer release()
+	var record, err = f.Await(c.Context) // Confused because record is wrong type, so how get chan capability from record?
```
Comment (lthibault) [inline main.go]: For the avoidance of doubt:  you're getting the chan capability from the call to `n.Vat.Connect`.  You can't get it from a `routing.Record`.

File: main.go [hunk 4]
```diff
@@ -160,13 +177,23 @@ func SlightEchoHandler(writer http.ResponseWriter, request *http.Request) {
...
 
-func runWorker(c *cli.Context, n *server.Node) error {
-	time.Sleep(10 * time.Second)
-	return errors.New("WORKER NOT IMPLEMENTED")
+func runWorker(c *cli.Context, n *server.Node, g peer.ID) error {
+	fmt.Println("Sleeping before run worker...")
+	time.Sleep(5 * time.Second)
+	fmt.Println("Running worker...")
+
+	conn, err := n.Vat.Connect(c.Context, peer.AddrInfo{ID: g}, casm.BasicCap{"lb/chan", "lb/chan/packed"})
```
Comment (evan-schott) [inline main.go:185]: Is this line correct?

**Comment by evan-schott** [comment]
@lthibault was trying to configure file so can use debugger easily, but probably better to just use print/log statements anyway.

Anyways I think this gets it through first 4 bullets of #2

File: main.go (hunk 4 above)
Comment (lthibault) [reply main.go:185]: It looks right.  Are you hitting an error?

(you can omit the `"lb/chan/packed" if you want.)

File: main.go [hunk 5]
```diff
@@ -61,6 +61,11 @@ var flags = []cli.Flag{
 		Usage: "number of expected peers in the cluster",
 		Value: 2,
 	},
+	&cli.IntFlag{ // TODO: Probably remove this once done debugging
```
Comment (lthibault) [inline main.go:64]: Consider making this a `BoolFlag`, which you can set to `true` just by passing `-foo` with no arguments.

**Review by lthibault — APPROVED** [review]
I don't see any obvious issues.  If this works on your end, I think you can merge it.

See inline for a few small recommendations.

File: main.go (hunk 4 above)
Comment (evan-schott) [reply main.go:185]: I think I figured out the error. I was checking if err == nil instead of if err != nil down below :/

**Pull request merged**